### DISCUSSION TITLE
nextjs api on aws deployment

### DISCUSSION THREAD

**Discussion by degenSumi** [vercel/next.js#51020] [Help]
### Summary

To what i should change the path of the serverless endpoint created with nextjs when i deploy it to aws
I mean http://localhost:3000/api/.. to what i should replace it with while consuming the serverless api.

**Comment by devjiwonchoi** [comment]
You can use the environment variables.

## Environment Variables in Next.js

- Configure the endpoint on the `.env` files.
```
// .env.develop

API_ENDPOINT=http://localhost:3000/api
```

```js
// some page.tsx

// ...

const data = await fetch(`${process.env.API_ENDPOINT}/something`)

// ...
```

Note that the fetch URL is in a string with [Template Literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals).

By doing this you can differentiate between Develop, Production, and Local API_ENDPOINT by separating files as `.env.development`, `.env.production`, `.env.local`.

Note that the `.env` files should remain at the root of the project.

> If you are using a [/src directory](https://nextjs.org/docs/app/building-your-application/configuring/src-directory), env.* files should remain in the root of your project.

## Environment Variables in AWS

You can do the same in AWS. For example, AWS Amplify has its own section for setting Environment Variables in the advanced settings for the deployment.

Example: 

![Screenshot 2023-06-09 at 8 10 16 PM](https://github.com/vercel/next.js/assets/120007119/6e563bab-e79c-4950-aa3f-c05517277b8d)

**Set the `Key` as API_ENDPOINT, and `Value` as `http://example.com/api`**

In your case, you should know the `hostname` of your instance and use that in the Environment Variables.
It will be much easier if you have a custom domain like `example.com`.

Note that there is no **Automatic** way to set an environment variable on the server regardless of AWS or not due to security issues.

## Conclusion
Use `environment variables` in your project and set them on the AWS Environment Variables section of your running service.

## Reference
- [Next.js Environment Variables](https://nextjs.org/docs/app/building-your-application/configuring/environment-variables)
- [AWS Amplify Environment Variables](https://docs.aws.amazon.com/amplify/latest/userguide/environment-variables.html)
- [Template Literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)
